crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,17 +1,130 @@
+                    f,
+                    [group(&format_args![
+                        // A single element tuple always needs parentheses
+                        &text("("),
+                        soft_block_indent(&format_args![single.format(), &text(",")]),
+                        &text(")"),
+                    ])]
+                );
+            }
+            [.., last] => last,
+        };
+
+        let magic_trailing_comma = USE_MAGIC_TRAILING_COMMA
+            && first_non_trivia_token(last.range().end(), f.context().contents())
+                .map(|token| token.kind)
+                == Some(TokenKind::Comma);
+
+        if magic_trailing_comma {
+            // A magic trailing comma forces us to print in expanded mode since we have more than
+            // one element
+            write!(
+                f,
+                [
+                    // An expanded group always needs parentheses
+                    &text("("),
+                    hard_line_break(),
+                    block_indent(&ExprSequence::new(elts)),
+                    hard_line_break(),
+                    &text(")"),
+                ]
+            )?;
+        } else if is_parenthesized(*range, elts, f) {
+            // If the tuple has parentheses, keep them. Note that unlike other expr parentheses,
+            // those are actually part of the range
+            write!(
+                f,
+                [group(&format_args![
+                    // If it was previously parenthesized, add them again
+                    &text("("),
+                    soft_block_indent(&ExprSequence::new(elts)),
+                    &text(")"),
+                ])]
+            )?;
+        } else {
+            write!(
+                f,
+                [group(&format_args![
+                    // If there were previously no parentheses, add them only if the group breaks
+                    if_group_breaks(&text("(")),
+                    soft_block_indent(&ExprSequence::new(elts)),
+                    if_group_breaks(&text(")")),
+                ])]
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn fmt_dangling_comments(&self, _node: &ExprTuple, _f: &mut PyFormatter) -> FormatResult<()> {
+        // Handled in `fmt_fields`
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ExprSequence<'a> {
+    elts: &'a [Expr],
+}
+
+impl<'a> ExprSequence<'a> {
+    const fn new(elts: &'a [Expr]) -> Self {
+        Self { elts }
+    }
+}
+
+impl Format<PyFormatContext<'_>> for ExprSequence<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        f.join_with(&format_args!(text(","), soft_line_break_or_space()))
+            .entries(self.elts.iter().formatted())
+            .finish()?;
+        // We need a trailing comma on the last entry of an expanded group since we have more
+        // than one element
+        write!(f, [if_group_breaks(&text(",")), soft_line_break()])
```

Comment:
```suggestion
        write!(f, [if_group_breaks(&text(","))])
```

Adding a trailing line break should be handled by the code wrapping the sequence. I assume that's already the case because you always use `block_indent` and `soft_block_indent`